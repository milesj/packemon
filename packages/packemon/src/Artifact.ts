import execa from 'execa';
import glob from 'fast-glob';
import fs from 'fs-extra';
import { rollup } from 'rollup';
import { applyStyle } from '@boost/cli';
import { isObject, Path, toArray, VirtualPath } from '@boost/common';
import { createDebugger, Debugger } from '@boost/debug';
import { removeSourcePath } from './helpers/removeSourcePath';
import type { Package } from './Package';
import { getRollupConfig } from './rollup/config';
import type {
	ApiType,
	ArtifactState,
	Build,
	BuildOptions,
	BuildResult,
	BuildResultFiles,
	ConfigFile,
	FeatureFlags,
	Format,
	InputMap,
	PackageExportPaths,
	PackageExports,
	PackemonPackageFeatures,
	Platform,
	Support,
} from './types';

export class Artifact {
	api: ApiType = 'private';

	bundle: boolean = true;

	// List of artifacts to build
	builds: Build[] = [];

	// Stats on the built artifacts
	buildResult: BuildResult = { files: [], time: 0 };

	// Config object in which inputs are grouped in
	configGroup: number = 0;

	// List of custom Rollup externals
	externals: string[] = [];

	// Features unique to this artifact
	features: PackemonPackageFeatures = {};

	// Mapping of output names to input paths
	inputs: InputMap = {};

	// Namespace for UMD bundles
	namespace: string = '';

	// Platform code will run on
	platform: Platform = 'node';

	// The package being built
	package: Package;

	// Are multiple builds writing to the lib folder
	sharedLib: boolean = false;

	// Target version code will run in
	support: Support = 'stable';

	// Current state
	state: ArtifactState = 'pending';

	protected debug!: Debugger;

	constructor(pkg: Package, builds: Build[]) {
		this.package = pkg;
		this.builds = builds;
		this.debug = createDebugger(['packemon', 'artifact', pkg.getSlug(), this.getLabel()]);
	}

	/**
	 * Build code and types in parallel.
	 */
	async build(
		options: BuildOptions,
		features: FeatureFlags,
		packemonConfig: ConfigFile,
	): Promise<void> {
		await Promise.all([this.buildCode(features, packemonConfig), this.buildTypes(features)]);
	}

	/**
	 * Build code artifacts using Rollup. We'll spin up a Rollup instance for each output,
	 * as Rollup will then generate all the necessary formats. For example:
	 * index -> cjs, lib.
	 */
	async buildCode(features: FeatureFlags, packemonConfig: ConfigFile): Promise<void> {
		this.debug('Building code artifacts with Rollup');

		const { output = [], ...input } = await getRollupConfig(this, features, packemonConfig);
		const bundle = await rollup({
			...input,
			onwarn: /* istanbul ignore next */ ({ id, loc = {}, message }) => {
				this.logWithSource(message, 'warn', {
					id: id && id !== loc.file ? id : undefined,
					output: this.package.getSlug(),
					sourceColumn: loc.column,
					sourceFile: loc.file,
					sourceLine: loc.line,
				});
			},
		});

		const files: BuildResultFiles[] = [];

		await Promise.all(
			toArray(output).map(async (out, index) => {
				const { originalFormat = 'lib', ...outOptions } = out;

				this.debug(' - Writing `%s` output', originalFormat);

				// While testing we want to avoid writing files,
				// so use generate() instead of write():
				// https://github.com/rollup/rollup/issues/4082
				const result =
					process.env.NODE_ENV === 'test' && !process.env.PACKEMON_TEST_WRITE
						? await bundle.generate(outOptions)
						: await bundle.write(outOptions);

				// Update build results and stats
				const bundledCode = result.output.reduce((code, chunk) => {
					if (chunk.type === 'chunk') {
						files.push({
							code: chunk.code,
							file: `${originalFormat}/${chunk.fileName}`,
						});

						return code + chunk.code;
					}

					return code;
				}, '');

				this.builds[index].stats = {
					size: Buffer.byteLength(bundledCode),
				};

				return result;
			}),
		);

		this.buildResult.files = files;
	}

	/**
	 * Build type artifacts using TypeScript. We'll need to spin up an individual process
	 * for each format, as they may require different declaration outputs or compiler options.
	 */
	async buildTypes(features: FeatureFlags): Promise<void> {
		if (!features.typescript) {
			return;
		}

		this.debug('Building types artifacts with TypeScript');

		await Promise.all(
			this.builds.map((build) => {
				if (!build.declaration) {
					return Promise.resolve();
				}

				const args: string[] = [];

				if (features.typescriptComposite) {
					args.push('--build', '--force', `tsconfig.${build.format}.json`);
				} else {
					args.push(
						'--declaration',
						'--declarationDir',
						build.format,
						'--declarationMap',
						'--emitDeclarationOnly',
						'--project',
						`tsconfig.${build.format}.json`,
					);
				}

				return execa('tsc', args, {
					cwd: this.package.path.path(),
					preferLocal: true,
				});
			}),
		);

		if (this.features.cjsTypesCompat) {
			const hasCjs = this.builds.some((build) => build.format === 'cjs');

			if (hasCjs) {
				this.debug('CJS types compatibility enabled, renaming `.d.ts` to `.d.cts`');

				const dtsFiles = await glob('**/*.d.ts', {
					absolute: true,
					cwd: this.package.path.append('cjs').path(),
				});

				await Promise.all(
					dtsFiles.map((dtsFile) => {
						const dtsPath = Path.create(dtsFile);
						const dtsName = dtsPath.name();

						return fs.rename(
							dtsPath.path(),
							dtsPath.parent().append(dtsName.replace('.d.ts', '.d.cts')).path(),
						);
					}),
				);
			}
		}
	}

	async clean() {
		this.debug('Cleaning artifact directories');

		const dirs = ['assets', 'dts', ...this.builds.map((build) => build.format)];

		await Promise.all(
			dirs.map(async (dir) => {
				const dirPath = this.package.path.append(dir).path();

				this.debug('  - %s', dirPath);

				await fs.remove(dirPath);
			}),
		);
	}

	findEntryPoint(formats: Format[], outputName: string) {
		for (const format of formats) {
			const build = this.builds.find((build) => build.format === format);

			if (build) {
				return this.getBuildOutput(format, outputName, build.declaration);
			}
		}

		return undefined;
	}

	// eslint-disable-next-line complexity
	getBuildOutput(format: Format, outputName: string, declaration: boolean = false) {
		const inputFile = this.inputs[outputName];
		const inputPath = inputFile ? removeSourcePath(inputFile) : undefined;
		let outputPath = outputName;

		// When using a public API, we do not create output files based on the input map.
		// Instead files mirror the source file structure, so we need to take that into account!
		if ((this.api === 'public' || !this.bundle) && inputPath) {
			outputPath = inputPath;
		}

		const folder = format === 'lib' && this.sharedLib ? `lib/${this.platform}` : format;
		// Folder path for declarations cannot have the platform subpath, since multiple builds can
		// share the same tsconfig
		const declFolder = format;
		const entryExt = format === 'cjs' || format === 'mjs' ? format : 'js';
		let declExt: string | undefined;

		if (declaration) {
			if (format === 'cjs' && this.features.cjsTypesCompat) {
				declExt = 'd.cts';
			} else if (!inputFile || /\.tsx?$/.test(inputFile)) {
				declExt = 'd.ts';
			} else if (inputFile.endsWith('.cts')) {
				declExt = 'd.cts';
			} else if (inputFile.endsWith('.mts')) {
				declExt = 'd.mts';
			}
		}

		return {
			declExt,
			declPath: declExt
				? `./${new VirtualPath(declFolder, `${inputPath ?? outputPath}.${declExt}`)}`
				: undefined,
			entryExt,
			entryPath: `./${new VirtualPath(folder, `${outputPath}.${entryExt}`)}`,
			folder,
		};
	}

	getIndexInput(): string {
		return this.inputs.index ? 'index' : Object.keys(this.inputs)[0];
	}

	getInputPaths(): InputMap {
		return Object.fromEntries(
			Object.entries(this.inputs).map(([outputName, inputFile]) => [
				outputName,
				// Return absolute paths so that Rollup paths/externals resolve correctly
				this.package.path.append(inputFile).path(),
			]),
		);
	}

	getLabel(): string {
		return `${this.platform}:${this.support}:${this.builds.map((build) => build.format).join(',')}`;
	}

	getPackageExports(features: FeatureFlags): PackageExports {
		const exportMap: PackageExports = {};

		if (this.api === 'private' || this.bundle) {
			Object.keys(this.inputs).forEach((outputName) => {
				this.mapPackageExportsFromBuilds(outputName, exportMap, features);
			});
		} else {
			// Use subpath export patterns when not bundling
			// https://nodejs.org/api/packages.html#subpath-patterns
			this.mapPackageExportsFromBuilds('*', exportMap, features);
			this.mapPackageExportsFromBuilds(this.getIndexInput(), exportMap, features, true);
		}

		return exportMap;
	}

	isComplete(): boolean {
		return this.state === 'passed' || this.state === 'failed';
	}

	isRunning(): boolean {
		return this.state === 'building';
	}

	toString(): string {
		return this.getLabel();
	}

	// eslint-disable-next-line complexity
	protected mapPackageExportsFromBuilds(
		outputName: string,
		exportMap: PackageExports,
		features: FeatureFlags,
		index: boolean = false,
	) {
		const defaultEntry = this.findEntryPoint(['lib'], outputName);
		let paths: PackageExportPaths = {};

		switch (this.platform) {
			case 'electron':
			case 'browser': {
				const esmEntry = this.findEntryPoint(['esm'], outputName);

				if (esmEntry) {
					paths = {
						types: esmEntry.declPath,
						module: esmEntry.entryPath, // Bundlers
						import: esmEntry.entryPath,
					};
				}

				// Node.js tooling
				const libEntry = this.findEntryPoint(['umd', 'lib'], outputName);

				paths.types ??= libEntry?.declPath;
				paths.default = libEntry?.entryPath;

				break;
			}

			case 'node': {
				const mjsEntry = this.findEntryPoint(['mjs'], outputName);
				const cjsEntry = this.findEntryPoint(['cjs'], outputName);

				if (mjsEntry && cjsEntry) {
					paths = {
						import: {
							types: mjsEntry.declPath,
							default: mjsEntry.entryPath,
						},
						require: {
							types: cjsEntry.declPath,
							default: cjsEntry.entryPath,
						},
					};
				} else if (mjsEntry) {
					paths = {
						types: mjsEntry.declPath,
						import: mjsEntry.entryPath,
					};
				} else if (cjsEntry) {
					paths = {
						types: cjsEntry.declPath,
						require: cjsEntry.entryPath,
					};

					// Automatically apply the mjs wrapper for cjs
					if (!paths.import && outputName !== '*') {
						paths.import = cjsEntry.entryPath.replace('.cjs', '-wrapper.mjs');
					}
				}

				if (defaultEntry) {
					if (!paths.types && !isObject(paths.import) && !isObject(paths.require)) {
						paths.types = defaultEntry.declPath;
					}

					if (!paths.require) {
						paths.default = defaultEntry.entryPath;
					}
				}

				break;
			}

			case 'native':
				paths.types = defaultEntry?.declPath;
				paths.default = defaultEntry?.entryPath;
				break;

			default:
				break;
		}

		const pathsMap: Record<string, PackageExportPaths | string> = {
			[this.platform === 'native' ? 'react-native' : this.platform]: paths,
		};

		// Point to the source files when using solid
		if (features.solid) {
			if (outputName === '*') {
				pathsMap.solid = `./src/*.${features.typescript ? 'tsx' : 'js'}`;
			} else {
				const input = this.inputs[outputName];

				pathsMap.solid = input.startsWith('./') ? input : `./${input}`;
			}
		}

		// Provide fallbacks if condition above is not
		if (defaultEntry) {
			pathsMap.default = defaultEntry.entryPath;
			pathsMap.types = defaultEntry.declPath!;
		}

		// eslint-disable-next-line no-param-reassign
		exportMap[index || outputName === 'index' ? '.' : `./${outputName}`] = pathsMap;
	}

	protected logWithSource(
		message: string,
		level: 'error' | 'info' | 'warn',
		{
			id,
			output,
			sourceColumn,
			sourceFile,
			sourceLine,
		}: {
			id?: string;
			output?: string;
			sourceColumn?: number;
			sourceFile?: string;
			sourceLine?: number;
		} = {},
	) {
		let msg = `[${this.package.getName()}${output ? `:${output}` : ''}] ${message}`;

		const meta: string[] = [];

		if (id) {
			meta.push(`id=${id}`);
		}

		if (sourceFile) {
			meta.push(
				`file=${new Path(sourceFile).path().replace(this.package.path.path(), '').slice(1)}`,
			);
		}

		if (sourceLine || sourceColumn) {
			meta.push(`line=${sourceLine ?? '?'}:${sourceColumn ?? '?'}`);
		}

		if (meta.length > 0) {
			msg += applyStyle(` (${meta.join(' ')})`, 'muted');
		}

		console[level](msg);
	}
}
