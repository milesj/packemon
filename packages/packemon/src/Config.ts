import { Blueprint, Schemas } from '@boost/common';
import { Configuration } from '@boost/config';
import { ConfigFile, ConfigMutator } from './types';

export class Config extends Configuration<ConfigFile> {
	blueprint({ func }: Schemas): Blueprint<ConfigFile> {
		return {
			babel: func(),
			rollup: func(),
		};
	}

	override bootstrap() {
		this.configureFinder({
			errorIfNoRootFound: false,
			extensions: ['js', 'ts'],
		});

		this.addProcessHandler('babel', this.wrapMutator);
		this.addProcessHandler('rollup', this.wrapMutator);
	}

	wrapMutator<T>(prev?: ConfigMutator<T>, next?: ConfigMutator<T>) {
		return (options: T) => {
			prev?.(options);
			next?.(options);
		};
	}
}
