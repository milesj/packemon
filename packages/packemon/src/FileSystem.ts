import fs from 'node:fs';
import { json } from '@boost/common';

export interface FileSystem {
	createDirAll: (path: string) => void;
	exists: (path: string) => boolean;
	readFile: (path: string) => string;
	readJson: <T>(path: string) => T;
	remove: (path: string) => void;
	writeFile: (path: string, data: string) => void;
	writeJson: (path: string, data: unknown) => void;
}

export const nodeFileSystem: FileSystem = {
	createDirAll: (path) => fs.mkdirSync(path, { recursive: true }),
	exists: (path) => fs.existsSync(path),
	readFile: (path) => fs.readFileSync(path, 'utf8'),
	readJson: (path) => json.load(path),
	remove: fs.unlinkSync,
	writeFile: (path, data) => void fs.writeFileSync(path, data, 'utf8'),
	writeJson: (path, data) =>
		void fs.writeFileSync(path, json.stringify(data, { space: 2 }), 'utf8'),
};
