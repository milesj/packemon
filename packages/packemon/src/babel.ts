import fs from 'fs-extra';
import { TransformOptions as ConfigStructure } from '@babel/core';
import { Path, toArray } from '@boost/common';
import { FeatureFlags } from './types';
import {
	CodeArtifact,
	DEFAULT_SUPPORT,
	Format,
	getBabelInputConfig,
	getBabelOutputConfig,
	Package,
	PackemonPackage,
	Platform,
	Project,
	Support,
} from '.';

const format = (process.env.PACKEMON_FORMAT ?? 'lib') as Format;
const support = (process.env.PACKEMON_SUPPORT ?? DEFAULT_SUPPORT) as Support;
const project = new Project(process.cwd());

function getBabelConfig(artifact: CodeArtifact, featureFlags: FeatureFlags): ConfigStructure {
	const inputConfig = getBabelInputConfig(artifact, featureFlags);
	const outputConfig = getBabelOutputConfig(
		artifact.platform,
		artifact.support,
		artifact.builds[0].format,
		featureFlags,
	);

	return {
		// Input must come first
		plugins: [...inputConfig.plugins!, ...outputConfig.plugins!],
		// Env must come first
		presets: [...outputConfig.presets!, ...inputConfig.presets!],
	};
}

export interface ConfigOptions {
	format?: Format;
	platform?: Platform;
	support?: Support;
}

export function createConfig(folder: string, options: ConfigOptions = {}): ConfigStructure {
	const path = new Path(folder);
	const contents = fs.readJsonSync(path.append('package.json').path()) as PackemonPackage;

	// Create package and configs
	const pkg = new Package(project, path, contents);

	if (pkg.packageJson.packemon) {
		pkg.setConfigs(toArray(pkg.packageJson.packemon));
	}

	// Determine the lowest platform to support
	const platforms = pkg.configs.map((config) => config.platform);
	let lowestPlatform: Platform = 'node';

	// istanbul ignore next
	if (platforms.includes('browser')) {
		lowestPlatform = 'browser';
	} else if (platforms.includes('native')) {
		lowestPlatform = 'native';
	}

	// Generate artifact and builds
	const artifact = new CodeArtifact(pkg, [{ format: options.format ?? format }]);
	artifact.bundle = false;
	artifact.platform = options.platform ?? lowestPlatform;
	artifact.support = options.support ?? support;

	return getBabelConfig(artifact, pkg.getFeatureFlags());
}

export function createRootConfig(options?: ConfigOptions): ConfigStructure {
	const config = createConfig(process.cwd(), options);

	return {
		...config,
		babelrc: true,
		babelrcRoots: project.getWorkspaceGlobs({ relative: true }),
		// Support React Native libraries by default
		overrides: [
			{
				presets: ['@babel/preset-flow'],
				test: /node_modules\/((jest-)?react-native|@react-native(-community)?)/iu,
			},
		],
	};
}
