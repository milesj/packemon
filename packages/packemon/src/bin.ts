import { checkPackageOutdated, Program } from '@boost/cli';
import { getVersion } from './helpers/getVersion';
import {
	BuildCommand,
	CleanCommand,
	InitCommand,
	PackCommand,
	ScaffoldCommand,
	ValidateCommand,
	WatchCommand,
} from '.';

let version = '0.0.0-internal';

try {
	version = getVersion();
} catch {
	// Ignore
}

async function run() {
	const program = new Program({
		bin: 'packemon',
		footer: 'Documentation: https://packemon.dev',
		name: 'Packemon',
		version,
	});

	program
		.categories({
			filter: 'Filtering',
		})
		.middleware(checkPackageOutdated('packemon', version))
		.register(new BuildCommand())
		.register(new CleanCommand())
		.register(new InitCommand())
		.register(new PackCommand())
		.register(new ScaffoldCommand())
		.register(new ValidateCommand())
		.register(new WatchCommand());

	await program.runAndExit(process.argv);
}

void run();
