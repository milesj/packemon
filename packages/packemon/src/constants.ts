import { StyleType } from '@boost/cli';
import {
	ArtifactState,
	BrowserFormat,
	Format,
	NativeFormat,
	NodeFormat,
	Platform,
	Support,
} from './types';

export const ASSETS = [
	// Styles
	'.css',
	'.scss',
	'.sass',
	'.less',
	// Images
	'.svg',
	'.png',
	'.jpg',
	'.jpeg',
	'.gif',
	'.webp',
	// Audio
	'.ogg',
	'.mp3',
	'.mpe',
	'.mpeg',
	'.wav',
	// Video
	'.mp4',
	'.mov',
	'.avi',
	'.webm',
	// Fonts
	'.woff',
	'.woff2',
	'.ttf',
];

export const EXTENSIONS = ['.ts', '.tsx', '.cts', '.mts', '.js', '.jsx', '.cjs', '.mjs'];

export const EXCLUDE = [
	'**/node_modules/**',
	'**/tests/**',
	'**/__fixtures__/**',
	'**/__mocks__/**',
	'**/__tests__/**',
	'**/*.(config|test|spec).*',
	'**/*-(test|spec).*',
];

// Rust (swc) uses regex patterns
export const EXCLUDE_RUST = [
	'node_modules',
	'tests',
	'__fixtures__',
	'__mocks__',
	'__tests__',
	'\\.(config|test|spec)\\.[a-z]+$',
	'\\.(test|spec)\\.[a-z]+$',
];

// https://reactnative.dev/docs/javascript-environment
// Based on browserslist: https://github.com/browserslist/browserslist
export const NATIVE_TARGETS: { [K in Support]: string } = {
	legacy: 'iOS 13', // 2019
	stable: 'iOS 14', // 2020
	current: 'iOS 15', // 2021
	experimental: 'iOS 16', // 2022
};

// Based on LTS schedule: https://nodejs.org/en/about/releases/
export const NODE_SUPPORTED_VERSIONS: { [K in Support]: string } = {
	legacy: '14.15.0', // LTS support started
	stable: '16.12.0', // ESM loader hooks
	current: '18.12.0', // LTS support started
	experimental: '19.0.0',
};

export const NPM_SUPPORTED_VERSIONS: { [K in Support]: string } = {
	legacy: '6.14.0',
	stable: '8.1.0',
	current: '8.19.0',
	experimental: '9.0.0',
};

// Based on browserslist: https://github.com/browserslist/browserslist
export const BROWSER_TARGETS: { [K in Support]: string[] | string } = {
	legacy: ['>=0.10%', 'not IE 11'],
	stable: ['defaults', 'not IE 11'],
	current: ['>=1%', 'not dead'],
	experimental: ['last 2 chrome versions', 'last 2 firefox versions'],
};

export const ELECTRON_TARGETS: { [K in Support]: string } = {
	legacy: '7', // Oct 2019
	stable: '11', // Nov 2020
	current: '16', // Nov 2021
	experimental: '21', // Sep 2022
};

export const SUPPORT_TO_ESM_SPEC = {
	legacy: 'es2018',
	stable: 'es2019',
	current: 'es2020',
	experimental: 'es2022',
} as const;

export const STATE_COLORS: { [K in ArtifactState]?: StyleType } = {
	pending: 'muted',
	passed: 'success',
	failed: 'failure',
};

export const DEFAULT_FORMATS: Record<Platform, Format> = {
	browser: 'esm',
	electron: 'esm',
	native: 'lib',
	node: 'mjs',
};

export const FORMATS_BROWSER: BrowserFormat[] = ['lib', 'esm', 'umd'];

export const FORMATS_ELECTRON: BrowserFormat[] = ['lib', 'esm'];

export const FORMATS_NATIVE: NativeFormat[] = ['lib'];

export const FORMATS_NODE: NodeFormat[] = ['lib', 'mjs', 'cjs'];

export const FORMATS: Format[] = ['lib', 'esm', 'umd', 'mjs', 'cjs'];

export const DEFAULT_INPUT = 'src/index.ts';

export const DEFAULT_PLATFORM: Platform = 'browser';

export const PLATFORMS: Platform[] = ['native', 'electron', 'node', 'browser'];

export const DEFAULT_SUPPORT: Support = 'stable';

export const SUPPORTS: Support[] = ['legacy', 'stable', 'current', 'experimental'];

export const SUPPORT_PRIORITY: Record<Support, number> = {
	legacy: 0,
	stable: 1,
	current: 2,
	experimental: 3,
};
