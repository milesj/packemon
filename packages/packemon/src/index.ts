/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './Artifact';
export * from './babel/config';
export * from './CodeArtifact';
export * from './commands/Base';
export * from './commands/Build';
export * from './commands/Clean';
export * from './commands/Init';
export * from './commands/Pack';
export * from './commands/Scaffold';
export * from './commands/Validate';
export * from './commands/Watch';
export * from './constants';
export * from './Package';
export * from './PackageValidator';
export * from './Packemon';
export * from './Project';
export * from './rollup/config';
export * from './types';
export * from './TypesArtifact';
