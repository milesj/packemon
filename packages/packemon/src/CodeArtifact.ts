import { rollup, RollupCache } from 'rollup';
import { Path, toArray } from '@boost/common';
import { createDebugger, Debugger } from '@boost/debug';
import { Artifact } from './Artifact';
import { removeSourcePath } from './helpers/removeSourcePath';
import { getRollupConfig } from './rollup/config';
import {
	BuildOptions,
	BuildResultFiles,
	CodeBuild,
	Format,
	InputMap,
	PackageExportPaths,
	PackageExports,
	Platform,
	Support,
} from './types';

export class CodeArtifact extends Artifact<CodeBuild> {
	bundle: boolean = true;

	cache?: RollupCache;

	// Config object in which inputs are grouped in
	configGroup: number = 0;

	// Mapping of output names to input paths
	inputs: InputMap = {};

	// Namespace for UMD bundles
	namespace: string = '';

	// Platform code will run on
	platform: Platform = 'node';

	// Are multiple builds writing to the lib folder
	sharedLib: boolean = false;

	// Target version code will run in
	support: Support = 'stable';

	protected debug!: Debugger;

	override startup() {
		this.debug = createDebugger(['packemon', 'code', this.package.getSlug(), this.getLabel()]);
	}

	override async cleanup(): Promise<void> {
		this.debug('Cleaning code artifacts');

		// Visualizer stats
		await this.removeFiles([this.package.project.root.append(this.getStatsFileName())]);
	}

	async build(options: BuildOptions): Promise<void> {
		this.debug('Building code artifacts with Rollup');

		const features = this.package.getFeatureFlags();

		if (options.analyze !== 'none') {
			features.analyze = options.analyze;
		}

		const { output = [], ...input } = getRollupConfig(this, features);
		const bundle = await rollup({
			...input,
			onwarn: /* istanbul ignore next */ ({ id, loc = {}, message }) => {
				this.logWithSource(message, 'warn', {
					id: id && id !== loc.file ? id : undefined,
					output: this.package.getSlug(),
					sourceColumn: loc.column,
					sourceFile: loc.file,
					sourceLine: loc.line,
				});
			},
		});

		if (bundle.cache) {
			this.cache = bundle.cache;
		}

		const files: BuildResultFiles[] = [];

		await Promise.all(
			toArray(output).map(async (out, index) => {
				const { originalFormat = 'lib', ...outOptions } = out;

				this.debug(' - Writing `%s` output', originalFormat);

				// While testing we want to avoid writing files,
				// so use generate() instead of write():
				// https://github.com/rollup/rollup/issues/4082
				const result =
					process.env.NODE_ENV === 'test'
						? await bundle.generate(outOptions)
						: await bundle.write(outOptions);

				// Update build results and stats
				const bundledCode = result.output.reduce((code, chunk) => {
					if (chunk.type === 'chunk') {
						files.push({
							code: chunk.code,
							file: `${originalFormat}/${chunk.fileName}`,
						});

						return code + chunk.code;
					}

					return code;
				}, '');

				this.builds[index].stats = {
					size: Buffer.byteLength(bundledCode),
				};

				return result;
			}),
		);

		this.buildResult.files = files;
	}

	findEntryPoint(formats: Format[], outputName: string): string {
		for (const format of formats) {
			if (this.builds.some((build) => build.format === format)) {
				return this.getBuildOutput(format, outputName).path;
			}
		}

		return '';
	}

	getBuildOutput(format: Format, outputName: string = '') {
		let name = outputName;

		// When not bundling, we do not create output files based on the input map.
		// Instead files mirror the source file structure, so we need to take that into account!
		if (!this.bundle && this.inputs[outputName]) {
			name = removeSourcePath(this.inputs[outputName]);
		}

		const ext = format === 'cjs' || format === 'mjs' ? format : 'js';
		const folder = format === 'lib' && this.sharedLib ? `lib/${this.platform}` : format;
		const file = `${name}.${ext}`;

		return {
			ext,
			file,
			folder,
			path: `./${new Path(folder, file)}`,
		};
	}

	getBuildTargets(): string[] {
		return this.builds.map((build) => build.format);
	}

	getInputPaths(): InputMap {
		// Return absolute paths so that Rollup paths/externals resolve correctly
		return Object.entries(this.inputs).reduce(
			(map, [outputName, inputFile]) => ({
				...map,
				[outputName]: this.package.path.append(inputFile).path(),
			}),
			{},
		);
	}

	getLabel(): string {
		return `${this.platform}:${this.support}:${this.getBuildTargets().join(',')}`;
	}

	getPackageExports(): PackageExports {
		const exportMap: PackageExports = {};

		Object.keys(this.inputs).forEach((outputName) => {
			const paths: PackageExportPaths = {};
			let libPath = '';

			this.builds.forEach(({ format }) => {
				const entry = this.findEntryPoint([format], outputName);

				switch (format) {
					case 'mjs':
					case 'esm':
						paths.import = entry;

						// Webpack and Rollup support
						if (format === 'esm') {
							paths.module = entry;
						}
						break;

					case 'cjs':
						paths.require = entry;
						break;

					case 'lib':
						libPath = entry;
						break;

					default:
						break;
				}
			});

			// Must come after import/require
			if (libPath) {
				paths.default = libPath;
			}

			exportMap[`./${outputName}`] = {
				[this.platform === 'native' ? 'react-native' : this.platform]:
					Object.keys(paths).length === 1 && libPath ? paths.default : paths,
			};
		});

		return exportMap;
	}

	getStatsFileName(): string {
		return `stats-${this.getStatsTitle().replace(/\//gu, '-')}.html`;
	}

	getStatsTitle(): string {
		return `${this.package.getName()}/${this.platform}/${this.support}`;
	}

	override toString() {
		return `code (${this.getLabel()})`;
	}
}
