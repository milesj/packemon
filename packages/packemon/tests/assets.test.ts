import fs from 'node:fs';
import { describe, expect, it, vi } from 'vitest';
import { Path } from '@boost/common';
import { Artifact } from '../src';
import { getFixturePath, loadPackageAtPath, snapshotPackageBuildOutputs } from './helpers';

vi.setConfig({ testTimeout: 30_000 });

describe('vanilla extract', () => {
	it('uses same assets across multiple formats', async () => {
		const root = new Path(getFixturePath('project-assets-vanilla'));
		const pkg = loadPackageAtPath(root);

		const index = new Artifact(pkg, [{ format: 'lib' }, { format: 'esm' }]);
		index.bundle = true;
		index.platform = 'browser';
		index.support = 'stable';
		index.inputs = { index: 'src/index.tsx' };

		pkg.artifacts.push(index);

		await pkg.build({ addEntries: false }, {});

		snapshotPackageBuildOutputs(pkg);
	});
});
