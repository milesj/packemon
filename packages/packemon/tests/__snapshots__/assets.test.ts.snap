// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vanilla extract > uses same assets across multiple formats > esm/index.js 1`] = `
"// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: stable, Format: esm

import { redText } from './style.css';
const MyComponent = props => /*#__PURE__*/React.createElement("div", Object.assign({
  className: redText
}, props));
export { MyComponent };
//# sourceMappingURL=index.js.map
"
`;

exports[`vanilla extract > uses same assets across multiple formats > lib/index.js 1`] = `
"// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: stable, Format: lib

'use strict';

const style_css = require('./style.css');
const MyComponent = props => /*#__PURE__*/React.createElement("div", Object.assign({
  className: style_css.redText
}, props));
exports.MyComponent = MyComponent;
//# sourceMappingURL=index.js.map
"
`;
